# My submission to the ArangoDb Hackathon
## By Patrick Wendo

The dataset used in this notebook is available on Kaggle [here](https://www.kaggle.com/datasets/andreuvallhernndez/myanimelist)

In [16]:
!pip3 install kagglehub pandas numpy networkx matplotlib nx_arangodb

Defaulting to user installation because normal site-packages is not writeable
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 67.8/67.8 KB 480.5 kB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 3.0 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.7/4.7 MB 3.6 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.5/116.5 KB 3.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.9/64.9 KB 3.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.2/18.2 MB 5.8 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.5/54.5 KB 2.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 146.1/146.1 KB 3.2 MB/s eta 0:00:0000:01
  Attempting uninstall: numpy
    Found existing installation: numpy 2.2.2
    Uninstalling numpy-2.2.2:
      Successfully uninstalled numpy-2.2.2
  Attempting uninstall: networkx
    Found existing installation: networkx 3.4.2
    Uninst

In [62]:
from arango import ArangoClient

import networkx as nx
import nx_arangodb as nxadb
import pandas as pd
import numpy as np
import networkx as nx
import matplotlib.pyplot as plt
from random import randint
import ast
import re

anime = pd.read_csv(
    "./datasets/myanimelist/anime.csv"
)

manga = pd.read_csv(
    "./datasets/myanimelist/manga.csv"
)

anime.info()
manga.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24985 entries, 0 to 24984
Data columns (total 39 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   anime_id          24985 non-null  int64  
 1   title             24985 non-null  object 
 2   type              24927 non-null  object 
 3   score             16050 non-null  float64
 4   scored_by         24985 non-null  int64  
 5   status            24985 non-null  object 
 6   episodes          24438 non-null  float64
 7   start_date        24110 non-null  object 
 8   end_date          22215 non-null  object 
 9   source            21424 non-null  object 
 10  members           24985 non-null  int64  
 11  favorites         24985 non-null  int64  
 12  episode_duration  24387 non-null  object 
 13  total_duration    24162 non-null  object 
 14  rating            24405 non-null  object 
 15  sfw               24985 non-null  bool   
 16  approved          24985 non-null  bool  

manga_id                                            title   type  score  \
0         2                                          Berserk  manga   9.47   
1        13                                        One Piece  manga   9.22   
2      1706  JoJo no Kimyou na Bouken Part 7: Steel Ball Run  manga   9.30   
3      4632                                   Oyasumi Punpun  manga   9.02   
4        25                              Fullmetal Alchemist  manga   9.03   

   scored_by                status  volumes  chapters  start_date    end_date  \
0     319696  currently_publishing      NaN       NaN  1989-08-25         NaN   
1     355375  currently_publishing      NaN       NaN  1997-07-22         NaN   
2     151433              finished     24.0      96.0  2004-01-19  2011-04-19   
3     168459              finished     13.0     147.0  2007-03-15  2013-11-02   
4     153151              finished     27.0     116.0  2001-07-12  2010-09-11   

   members  favorites   sfw  approved          created_at_before  \
0   643969     119470  True      True  2007-07-17 20:14:45+00:00   
1   579557     111462  True      True  2007-07-17 20:14:45+00:00   
2   248511      41713  True      True  2007-10-07 08:14:20+00:00   
3   413897      49361  True      True  2008-02-03 15:54:30+00:00   
4   284027      29634  True      True  2007-07-17 20:14:45+00:00   

                  updated_at real_start_date real_end_date  \
0  2023-04-01 00:19:31+00:00      1989-08-25           NaN   
1  2023-06-24 12:39:48+00:00      1997-07-22           NaN   
2  2023-04-02 18:07:42+00:00      2004-01-19    2011-04-19   
3  2023-04-02 18:09:11+00:00      2007-03-15    2013-11-02   
4  2023-05-27 17:21:27+00:00      2001-07-12    2010-09-11   

                                              genres  \
0  ['Action', 'Adventure', 'Award Winning', 'Dram...   
1                 ['Action', 'Adventure', 'Fantasy']   
2  ['Action', 'Adventure', 'Mystery', 'Supernatur...   
3                         ['Drama', 'Slice of Life']   
4  ['Action', 'Adventure', 'Award Winning', 'Dram...   

                                              themes           demographics  \
0  ['Gore', 'Military', 'Mythology', 'Psychologic...             ['Seinen']   
1                                                 []            ['Shounen']   
2                                     ['Historical']  ['Seinen', 'Shounen']   
3                                  ['Psychological']             ['Seinen']   
4                                       ['Military']            ['Shounen']   

                                             authors  \
0  [{'id': 1868, 'first_name': 'Kentarou', 'last_...   
1  [{'id': 1881, 'first_name': 'Eiichiro', 'last_...   
2  [{'id': 2619, 'first_name': 'Hirohiko', 'last_...   
3  [{'id': 2836, 'first_name': 'Inio', 'last_name...   
4  [{'id': 1874, 'first_name': 'Hiromu', 'last_na...   

              serializations  \
0           ['Young Animal']   
1  ['Shounen Jump (Weekly)']   
2             ['Ultra Jump']   
3      ['Big Comic Spirits']   
4         ['Shounen Gangan']   

                                            synopsis  \
0  Guts, a former mercenary now known as the "Bla...   
1  Gol D. Roger, a man referred to as the "King o...   
2  In the American Old West, the world's greatest...   
3  Punpun Onodera is a normal 11-year-old boy liv...   
4  Alchemists are knowledgeable and naturally tal...   

                                          background  \
0  Berserk won the Award for Excellence at the si...   
1  One Piece is the highest selling manga series ...   
2  JoJo no Kimyou na Bouken Part 7: Steel Ball Ru...   
3  Oyasumi Punpun was originally serialized in Yo...   
4  Hagane no Renkinjutsushi won the 49th Shogakuk...   

                                        main_picture  \
0  https://cdn.myanimelist.net/images/manga/1/157...   
1  https://cdn.myanimelist.net/images/manga/2/253...   
2  https://cdn.myanimelist.net/images/manga/3/179...   
3  https://cdn.myanimelis

### Data Exploration

- We have 2 datasets, anime and manga. 
- These are both relational datasets with a single entry having multiple values. We need to figure out how to represent this data as a graph. 

#### Our Strategy.
- We could represent this as an attributed graph with node attributes and edge attributes. An **attributed graph** is one where aside from node labels, or edge labels, a node or an edge will also have additional metadata. For example, in a graph about people, we could have a node with the label "Anna" with attributes `{position: "CEO", start_date: "2019-08-19"}`. 

- The node label would be the name of the anime/manga. We could also have some columns be extracted to nodes of their own. For instance, an anime could fall into multiple genres. We could have each of those as a separate node. Similarly for columns like studios, themes, producers, licensors and demographics.

In [29]:
# Exploration of the anime dataset
pd.set_option("display.max.columns", None)
[anime.shape, manga.shape]

[(24985, 39), (64833, 30)]

- The anime dataset has about **24,985 rows with 39 columns**, while the manga dataset has **64,833 rows and 30 columns**

In [28]:
anime.head()

anime_id                               title   type  score  scored_by  \
0      5114    Fullmetal Alchemist: Brotherhood     tv   9.10    2037075   
1     11061              Hunter x Hunter (2011)     tv   9.04    1671587   
2     38524  Shingeki no Kyojin Season 3 Part 2     tv   9.05    1491491   
3      9253                         Steins;Gate     tv   9.07    1348232   
4     28851                      Koe no Katachi  movie   8.94    1540277   

            status  episodes  start_date    end_date        source  members  \
0  finished_airing      64.0  2009-04-05  2010-07-04         manga  3206028   
1  finished_airing     148.0  2011-10-02  2014-09-24         manga  2688079   
2  finished_airing      10.0  2019-04-29  2019-07-01         manga  2133927   
3  finished_airing      24.0  2011-04-06  2011-09-14  visual_novel  2463954   
4  finished_airing       1.0  2016-09-17  2016-09-17         manga  2218467   

   favorites episode_duration   total_duration rating   sfw  approved  \
0     219036  0 days 00:24:20  1 days 01:57:20      r  True      True   
1     202109  0 days 00:23:37  2 days 10:15:16  pg_13  True      True   
2      55644  0 days 00:23:55  0 days 03:59:10      r  True      True   
3     184312  0 days 00:24:20  0 days 09:44:00  pg_13  True      True   
4      84124  0 days 02:10:03  0 days 02:10:03  pg_13  True      True   

                  created_at                 updated_at  start_year  \
0  2008-08-21 03:35:22+00:00  2023-04-02 18:07:03+00:00      2009.0   
1  2011-07-27 08:41:39+00:00  2023-04-02 18:11:36+00:00      2011.0   
2  2018-10-14 18:13:45+00:00  2023-04-02 18:10:19+00:00      2019.0   
3  2010-07-26 09:23:40+00:00  2023-04-02 18:08:42+00:00      2011.0   
4  2014-12-15 08:00:46+00:00  2023-04-02 18:15:50+00:00      2016.0   

  start_season real_start_date real_end_date broadcast_day broadcast_time  \
0       spring      2009-04-05    2010-07-04        sunday       17:00:00   
1         fall      2011-10-02    2014-09-24        sunday       10:55:00   
2       spring      2019-04-29    2019-07-01        monday       00:10:00   
3       spring      2011-04-06    2011-09-14     wednesday       02:05:00   
4       summer      2016-09-17    2016-09-17           NaN            NaN   

                                        genres  \
0  ['Action', 'Adventure', 'Drama', 'Fantasy']   
1           ['Action', 'Adventure', 'Fantasy']   
2                          ['Action', 'Drama']   
3              ['Drama', 'Sci-Fi', 'Suspense']   
4                   ['Award Winning', 'Drama']   

                             themes demographics              studios  \
0                      ['Military']  ['Shounen']            ['Bones']   
1                                []  ['Shounen']         ['Madhouse']   
2  ['Gore', 'Military', 'Survival']  ['Shounen']       ['Wit Studio']   
3  ['Psychological', 'Time Travel']           []        ['White Fox']   
4              ['Romantic Subtext']  ['Shounen']  ['Kyoto Animation']   

                                           producers  \
0  ['Aniplex', 'Square Enix', 'Mainichi Broadcast...   
1   ['VAP', 'Nippon Television Network', 'Shueisha']   
2  ['Production I.G', 'Dentsu', 'Mainichi Broadca...   
3  ['Frontier Works', 'Media Factory', 'Kadokawa ...   
4  ['Shochiku', 'Pony Canyon', 'Kodansha', 'ABC A...   

                              licensors  \
0  ['Funimation', 'Aniplex of America']   
1                         ['VIZ Media']   
2                        ['Funimation']   
3                        ['Funimation']   
4          ['Eleven Arts', 'NYAV Post']   

                                            synopsis  \
0  After a horrific alchemy experiment goes wrong...   
1  Hunters devote themselves to accomplishing haz...   
2  Seeking to restore humanity's diminishing hope...   
3  Eccentric scientist Rintarou Okabe has a never...   
4  As a wild youth, elementary school student Sho...   

                                          background  \
0            

In [32]:
manga.head()

manga_id                                            title   type  score  \
0         2                                          Berserk  manga   9.47   
1        13                                        One Piece  manga   9.22   
2      1706  JoJo no Kimyou na Bouken Part 7: Steel Ball Run  manga   9.30   
3      4632                                   Oyasumi Punpun  manga   9.02   
4        25                              Fullmetal Alchemist  manga   9.03   

   scored_by                status  volumes  chapters  start_date    end_date  \
0     319696  currently_publishing      NaN       NaN  1989-08-25         NaN   
1     355375  currently_publishing      NaN       NaN  1997-07-22         NaN   
2     151433              finished     24.0      96.0  2004-01-19  2011-04-19   
3     168459              finished     13.0     147.0  2007-03-15  2013-11-02   
4     153151              finished     27.0     116.0  2001-07-12  2010-09-11   

   members  favorites   sfw  approved          created_at_before  \
0   643969     119470  True      True  2007-07-17 20:14:45+00:00   
1   579557     111462  True      True  2007-07-17 20:14:45+00:00   
2   248511      41713  True      True  2007-10-07 08:14:20+00:00   
3   413897      49361  True      True  2008-02-03 15:54:30+00:00   
4   284027      29634  True      True  2007-07-17 20:14:45+00:00   

                  updated_at real_start_date real_end_date  \
0  2023-04-01 00:19:31+00:00      1989-08-25           NaN   
1  2023-06-24 12:39:48+00:00      1997-07-22           NaN   
2  2023-04-02 18:07:42+00:00      2004-01-19    2011-04-19   
3  2023-04-02 18:09:11+00:00      2007-03-15    2013-11-02   
4  2023-05-27 17:21:27+00:00      2001-07-12    2010-09-11   

                                              genres  \
0  ['Action', 'Adventure', 'Award Winning', 'Dram...   
1                 ['Action', 'Adventure', 'Fantasy']   
2  ['Action', 'Adventure', 'Mystery', 'Supernatur...   
3                         ['Drama', 'Slice of Life']   
4  ['Action', 'Adventure', 'Award Winning', 'Dram...   

                                              themes           demographics  \
0  ['Gore', 'Military', 'Mythology', 'Psychologic...             ['Seinen']   
1                                                 []            ['Shounen']   
2                                     ['Historical']  ['Seinen', 'Shounen']   
3                                  ['Psychological']             ['Seinen']   
4                                       ['Military']            ['Shounen']   

                                             authors  \
0  [{'id': 1868, 'first_name': 'Kentarou', 'last_...   
1  [{'id': 1881, 'first_name': 'Eiichiro', 'last_...   
2  [{'id': 2619, 'first_name': 'Hirohiko', 'last_...   
3  [{'id': 2836, 'first_name': 'Inio', 'last_name...   
4  [{'id': 1874, 'first_name': 'Hiromu', 'last_na...   

              serializations  \
0           ['Young Animal']   
1  ['Shounen Jump (Weekly)']   
2             ['Ultra Jump']   
3      ['Big Comic Spirits']   
4         ['Shounen Gangan']   

                                            synopsis  \
0  Guts, a former mercenary now known as the "Bla...   
1  Gol D. Roger, a man referred to as the "King o...   
2  In the American Old West, the world's greatest...   
3  Punpun Onodera is a normal 11-year-old boy liv...   
4  Alchemists are knowledgeable and naturally tal...   

                                          background  \
0  Berserk won the Award for Excellence at the si...   
1  One Piece is the highest selling manga series ...   
2  JoJo no Kimyou na Bouken Part 7: Steel Ball Ru...   
3  Oyasumi Punpun was originally serialized in Yo...   
4  Hagane no Renkinjutsushi won the 49th Shogakuk...   

                                        main_picture  \
0  https://cdn.myanimelist.net/images/manga/1/157...   
1  https://cdn.myanimelist.net/images/manga/2/253...   
2  https://cdn.myanimelist.net/images/manga/3/179...   
3  https://cdn.myanimelis

##### Data Fix #1
- Fix string that look like lists, `"[list, item]" -> ["list", "item"]`. This will allow us to use the #explode() function later when creating data in NetworkX

In [119]:
anime_columns_to_fix = ["genres", "themes", "demographics", "title_synonyms", "studios"]
manga_columns_to_fix = ["genres", "themes", "demographics", "authors"]
print("Initial Data anime Type: \n", anime.loc[1:2]['title_synonyms'].apply(type))
print("Initial Data manga Type: \n", manga.loc[1:2]['authors'].apply(type))

def rewrite_anime(col):
    anime[col] = anime[col].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else x)

def rewrite_manga(col):
    manga[col] = manga[col].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else x)

for col in anime_columns_to_fix:
    rewrite_anime(col)

for col in manga_columns_to_fix:
    rewrite_manga(col)

print("After Fix anime Data: \n", anime.loc[1:2]['title_synonyms'].apply(type))
print("After Fix manga Data: \n", manga.loc[1:2]['authors'].apply(type))

Initial Data anime Type: 
 1    <class 'list'>
2    <class 'list'>
Name: title_synonyms, dtype: object
Initial Data manga Type: 
 1    <class 'list'>
2    <class 'list'>
Name: authors, dtype: object
After Fix anime Data: 
 1    <class 'list'>
2    <class 'list'>
Name: title_synonyms, dtype: object
After Fix manga Data: 
 1    <class 'list'>
2    <class 'list'>
Name: authors, dtype: object


In [104]:
anime_subset = anime.loc[0:5]
exploded = anime_subset.explode('themes') # fails because genres is a string
exploded['themes']


0            Military
1                 NaN
2                Gore
2            Military
2            Survival
3       Psychological
3         Time Travel
4    Romantic Subtext
5    Romantic Subtext
Name: themes, dtype: object

In [ ]:
## Identify Key groups in the following columns: "genres", "themes", "demographics", "studios". These could be shared between anime entries.

genre_explode = anime.explode('genres')
genre_grouped_data = genre_explode.groupby("genres")
anime_genres = genre_grouped_data.groups.keys()
print("Genres: ", list(set(list(genre_grouped_data.groups.keys()))))

theme_explode = anime.explode('themes')
theme_grouped_data = theme_explode.groupby("themes")
anime_themes = theme_grouped_data.groups.keys()
print("Themes: ", list(set(list(theme_grouped_data.groups.keys()))))

demographics_explode = anime.explode('demographics')
demographic_grouped_data = demographics_explode.groupby("demographics")
anime_demographics = demographic_grouped_data.groups.keys()
print("Demographics: ", list(set(list(demographic_grouped_data.groups.keys()))))

studios_explode = anime.explode('studios')
studio_grouped_data = studios_explode.groupby("studios")
anime_studios = studio_grouped_data.groups.keys()
print("Studios: ", list(set(list(studio_grouped_data.groups.keys()))))

Genres:  ['Ecchi', 'Slice of Life', 'Action', 'Sci-Fi', 'Adventure', 'Horror', 'Suspense', 'Award Winning', 'Comedy', 'Avant Garde', 'Boys Love', 'Drama', 'Sports', 'Erotica', 'Girls Love', 'Fantasy', 'Supernatural', 'Romance', 'Mystery', 'Hentai', 'Gourmet']
Themes:  ['Anthropomorphic', 'Martial Arts', 'Music', 'Childcare', 'Historical', 'Parody', 'Team Sports', 'Vampire', 'Space', 'Harem', 'Mythology', 'Crossdressing', 'CGDCT', 'Pets', 'Romantic Subtext', 'Workplace', 'Love Polygon', 'Gore', 'High Stakes Game', 'Mahou Shoujo', 'Reincarnation', 'Medical', 'Psychological', 'Delinquents', 'Racing', 'Iyashikei', 'Detective', 'Military', 'Strategy Game', 'Visual Arts', 'Time Travel', 'Idols (Male)', 'Mecha', 'Survival', 'Samurai', 'School', 'Performing Arts', 'Educational', 'Isekai', 'Gag Humor', 'Otaku Culture', 'Super Power', 'Organized Crime', 'Reverse Harem', 'Video Game', 'Combat Sports', 'Showbiz', 'Magical Sex Shift', 'Idols (Female)', 'Adult Cast']
Demographics:  ['Seinen', 'Kids'

##### Authors
- The author field in Manga datasets should be extracted into it's own dataset. Each author will be a node on it's own. They may be shared across datasets. The node label will be the id with attributes being first name, last name and role. 

In [124]:
manga_authors = manga["authors"]
manga_authors.head()[0]

[{'id': 1868,
  'first_name': 'Kentarou',
  'last_name': 'Miura',
  'role': 'Story & Art'},
 {'id': 49592, 'first_name': '', 'last_name': 'Studio Gaga', 'role': 'Art'}]